### PR TITLE
Implement Google Picker support in the new file picker

### PR DESCRIPTION
**Pull request by robertknight**
This PR re-implements support for choosing files from Google Drive in the new file picker. Rather than re-use the existing code I've taken the opportunity to factor out the Google Picker client into a separate module.

Changes compared to the previous implementation:

- The "Select from Google" button appears immediately when the picker form loads, instead of appearing only after the Google API client libraries have loaded. Having existing controls on a form jump around as functionality loads is really irritating (👀 looking at you Trello!).  Instead if the user clicks the button before the functionality has loaded, we'll just show a "waiting" indicator.
- I've removed some code from the previous implementation that dealt with Google Docs, but was never finished

----

**Remaining tasks for this PR**:

- [x] Show a loading indicator when the "Select from Google Drive" button is clicked if the picker doesn't load immediately
- [x] Prevent the user from choosing a file from an alternative source if they are in the middle of picking a file from Google Drive
- [x] Show a loading indicator after the user makes a selection while the sharing settings are being updated
- [x] Hide the loading indicator if the user cancels authorization or choosing a file
- [x] Show an error indicator if something goes wrong when loading the Google libraries, authorizing or changing sharing settings
- [x] Figure out why the file picker works in Chrome (on my system) but not Safari (this was a bug with the old implementation as well, see https://github.com/hypothesis/lms/issues/262) _(Works in Chrome, Firefox Stable and Safari Tech Preview on my system, but not Safari stable. Will look into that separately)_
- [x] Tests for changes in the `FilePickerApp`
- [x] Tests for the Google Picker client
- [x] Tests for the error dialog

**Outside the scope of this PR, but we should revisit**:

- Show a more explicit indication that the sharing settings on the file were changed

### CODE REVIEW

**Comment by robertknight** [comment]
I can get a little further in the process with the current Safari Tech Preview (Release 83 (Safari 12.2, WebKit 14608.1.25)). After clicking "Select from Google Drive" and signing in to Google, the Google Drive picker appears. After selecting a file however, the loading indicator remains visible and the form never gets submitted.

**Comment by robertknight** [comment]
Hmm, I tried again and this time couldn't reproduce the issue in FF Nightly. 🤔 - More debugging to do next week.

**Comment by robertknight** [comment]
I've decided to look into the issue I see in the stable release of Safari separately, since this works in Chrome and Firefox Stable on my system, as well as the Safari Tech Preview.

File: lms/templates/content_item_selection/new_content_item_selection.html.jinja2 [hunk 1]
```diff
@@ -58,7 +58,10 @@
   "courseId": "{{ course_id }}",
   "formAction": "{{ content_item_return_url }}",
   "formFields": {{ form_fields | tojson }},
+  "googleClientId": "{{ google_client_id }}",
+  "googleDeveloperKey": "{{ google_developer_key }}",
   "lmsName" : "Canvas",
+  "lmsUrl": "{{ lms_url }}",
```
Comment (robertknight) [inline lms/templates/content_item_selection/new_content_item_selection.html.jinja2]: This is the URL (at least the origin, I think the path is optional) of the top-level frame in the file picker modal and is needed by the Google Picker for security reasons.

**Comment by seanh** [comment]
When selecting a Canvas file, when the authorization popup window is open, the dialog behind it changes to a view that explains what the user needs to do and includes an <kbd>Authorize</kbd> button that re-raises the popup window if the user has lost it, and a <kbd>Cancel</kbd> button that closes the popup and goes back to the previous view:

Peek 2019-06-07 12-47

Can the Google Picker should work in the same way? When the Google Picker authorization popup window is open show a *Allow file access* dialog (_To select a file, you must authorize Hypothesis to access your files in Google Drive_) with <kbd>Cancel</kbd> and <kbd>Authorize</kbd> buttons?

Current Google Picker behaviour:

Peek 2019-06-07 12-49

**Comment by seanh** [comment]
When the feature flag is turned off, the old Google Picker support seems to configure Google Picker to show **Google Drive** as the first tab and **Upload** as a secondary option:

Screenshot from 2019-06-07 12-57-50

But with the new picker the tabs are the other way round, to access files from Drive users will have to spot the **Google Drive** tab:

Screenshot from 2019-06-07 13-00-09

Can this be changed to make the **Google Drive** tab the default / first option?

**Comment by seanh** [comment]
The <kbd>Select PDF from Google Drive</kbd> button only seems to appear when I launch the app in Canvas but not when I launch it in Potemkin (and, I'm guessing, not in other non-Canvas LMS's like Moodle etc).

<kbd>Select PDF from Google Drive</kbd> should _always_ be shown -- it doesn't verify with the LMS or the application instance and there's nothing the user needs to do to enable it, it's just there all the time.

**Comment by seanh** [comment]
The **visit our support page** link tries to open <https://web.hypothes.is/help/> within the iframe but gets a "refused to connect" error from Chrome. I'm guessing because <https://web.hypothes.is/help/> refuses to be iframed. If I right-click and open the link in a new tab then it works.

**Comment by robertknight** [comment]
> The visit our support page link tries to open https://web.hypothes.is/help/ within the iframe but gets a "refused to connect" error from Chrome.

Oh, 🤦‍♂. It should have `target="_blank"`

**Comment by robertknight** [comment]
> The Select PDF from Google Drive button only seems to appear when I launch the app in Canvas but not when I launch it in Potemkin (and, I'm guessing, not in other non-Canvas LMS's like Moodle etc).

Have you configured Google API keys in Potemkin? It should always show up if the `GOOGLE_CLIENT_ID` and `GOOGLE_DEVELOPER_KEY` env vars are set.

**Comment by seanh** [comment]
Related to my earlier comment about whether the Google Picker can use an **Allow file access** dialog when the authorization popup window is open like the Canvas Picker does: once all the error-handling for the Canvas Picker is in place the intended behaviour of the Canvas Picker whenever something goes wrong (for example the file list proxy API responds with an error) is that **it should return to this view**:

Screenshot from 2019-06-07 13-57-30

As a bonus it might show an error message under the _To select a file, you must authorize Hypothesis to access your files in Canvas_ sentence (and this is where a link to our support page, and details of what went wrong, could go) but the main thing is that it:

1. Has an <kbd>Authorize</kbd> button so that the user can try starting again
2. Has a <kbd>Cancel</kbd> button so the user can go back and use a different type of file instead

The current behaviour when the Google Picker fails is to show this error dialog, which is different from the Canvas Picker's behaviour and lacks 1 and 2:

Screenshot from 2019-06-07 13-55-01

(The <kbd>X</kbd> on that dialog doesn't seem to work, but I think we should be getting rid of the dialog entirely anyway.)

**Comment by seanh** [comment]
> Have you configured Google API keys in Potemkin? It should always show up if the GOOGLE_CLIENT_ID and GOOGLE_DEVELOPER_KEY env vars are set.

These envvars belong to the LMS app itself, rather than to the LMS. I do have them configured and it works in Canvas but seems to not work in Potemkin. Just tested in Moodle and it's not working there either. This could be to do with that both Potemkin and Moodle use "unconfigured assignment" launches whereas Canvas uses a "content-item" launch.

**Pull request merged**